cilium: host proxy delegation infra

borkmann opened the pull request:
(see commit desc)

borkmann committed:
bpf: Move l7_lb_proxy_port into L7 helper

Move the check for the port into the lb{4,6}_svc_is_l7loadbalancer()
helper to make this consistent everywhere. Right now a port of 0 is
reserved/unused, so also add a comment to the code.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
bpf: Add a new punt to proxy drop code

Add a new drop code called DROP_PUNT_PROXY which can be used out of
lb{4,6}_local() to denote that the packet needs to be pushed up the
stack in order to delegate it to a proxy.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
bpf: Add proxy delegation flag

Add a new proxy delegation flag called SVC_FLAG_L7_DELEGATE. This
flag aliases to SVC_FLAG_LOOPBACK but does not overlap in its use.
The latter is only used in combination with HostPort, and the control
plane makes sure that there is no collision.

Reusing SVC_FLAG_L7_LOADBALANCER was not an option since this collides
with session affinity / per-service algorithm. For services marked
with SVC_FLAG_L7_DELEGATE, the behavior is that in lb{4,6}_local() we
check whether the selected backend is local on the node, and if it is,
then the packet is punted up to a user space proxy for processing. If
the backend is non-local, then the dataplane forwards the traffic as
per underlying KPR configuration to the remote backend.

This is different to SVC_FLAG_L7_LOADBALANCER since in that case all
traffic no matter the backend is punted up to a proxy. The new flag
SVC_FLAG_L7_DELEGATE therefore can help scalability e.g. going from
L3/4 up to L7 since the former can be handled in tcx/XDP.

For consistency, also do a few renames SVC_FLAG_L7LOADBALANCER into
SVC_FLAG_L7_LOADBALANCER, and *lb{4,6}_svc_is_l7loadbalancer() into
*lb{4,6}_svc_is_l7_loadbalancer().

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
cilium, docs: Add paragraph on host proxy delegation

Add related docs to the KPR guide for proxy delegation annotation along
with usage yaml example.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
cilium: Add ServiceProxyDelegation annotation

Add a new service.cilium.io/proxy-delegation annotation with possible
values of {delegate-if-local, none}. As explained in the previous commit
the idea of the delegate-if-local is that if a service backend is a
local backend (e.g. host address), then punt the packet up to the stack
in its original form so that a L7 proxy such as Envoy can take care of
the request. However, if the backend is a remote node or on a remote node
then the BPF datapath takes care of the forwarding based on the underlying
configured forwarding method.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
cilium: Wire-up ServiceProxyDelegation into service control plane

Wire-up to push the annotation all the way through so that relevant bits
are being set in the service map.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
cilium: Extend externalTrafficPolicy=Local for proxy delegation

Extend externalTrafficPolicy=Local to also support proxy delegation.
This can be useful for the case where all traffic should be pushed
up to a local proxy. This reuses the existing means to filter out
remote IPs and only have local proxy as backend, example:

  apiVersion: v1
  kind: Service
  metadata:
    name: example
    namespace: default
    annotations:
      service.cilium.io/type: LoadBalancer
      service.cilium.io/src-ranges-policy: allow
      service.cilium.io/proxy-delegation: delegate-if-local
  spec:
    type: LoadBalancer
    externalTrafficPolicy: Local
    loadBalancerSourceRanges:
      - 192.168.2.0/24
    ports:
      - name: http
        protocol: TCP
        port: 80
        targetPort: 80

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

aspsk commented on bpf/lib/lb.h:
```diff
@@ -368,9 +368,30 @@ bool lb6_svc_is_localredirect(const struct lb6_service *svc)
 #endif /* ENABLE_LOCAL_REDIRECT_POLICY */
 
+static __always_inline
+bool __lb4_svc_is_l7loadbalancer(const struct lb4_service *svc __maybe_unused)
```
Are these `__lb{4,6}_svc_is_l7loadbalancer` helpers ever used outside of corresponding `lb{4,6}_svc_is_l7loadbalancer` helpers? If not, then maybe omit them to save a few lines?

borkmann replied:
fixed it, actually for bind restriction it would make sense to use the __ helper.

aspsk reviewed (approved):
I have one comment, but maybe it is not worth re-running the CI

squeed commented on bpf/lib/nodeport.h:
```diff
@@ -1260,6 +1260,12 @@ static __always_inline int nodeport_svc_lb6(struct __ctx_buff *ctx,
 	ret = lb6_local(get_ct_map6(tuple), ctx, l3_off, l4_off,
 			key, tuple, svc, &ct_state_svc,
 			nodeport_xlate6(svc, tuple), ext_err, 0);
+#ifdef ENABLE_L7_LB
```
just a question for my own knowledge (feel free to resolve): Why do we need a new mechanism to redirect to the proxy? Why isn't the existing one sufficient?

borkmann replied:
The old one pushes all traffic into the proxy, but this is not what we want given only traffic should be pushed into a proxy iff there was a "backend" which is a local proxy, and if there is not, we push all traffic via xdp/bpf datapath off the node again. Meaning, these are two different mechanisms. Reusing the same flag with the very limited bits we have in the svc value is also tricky since proxy port overlaps with session affinity timeout, and we would like to use the latter feature too. Hope that helps.

squeed commented on pkg/annotation/k8s.go:
```diff
@@ -145,6 +145,12 @@ const (
+	// ServiceProxyDelegation is the annotation name used to specify whether there
+	// should be delegation to a 3rd party proxy. Allowed value is "delegate-if-local".
+	// The latter pushes all service packets to a user space proxy if the backend IP
+	// is the IP of the local node.
```
Does this mean that *all* SYNs are punted to L7 if *any* service endpoint is the local node IP? If so, it might be worth noting if this overrides traffic policy?

borkmann replied:
No, this is only if the BPF LB-algorithm selects a backend where the backend IP is a local IP. If it selects a remote IP, BPF datapath does DNAT/SNAT and pushes the packet back out.

borkmann replied:
I'll refine the wording a bit in the comment.

squeed commented on pkg/maps/lbmap/lbmap.go:
```diff
@@ -602,14 +602,23 @@ func (*LBBPFMap) IsMaglevLookupTableRecreated(ipv6 bool) bool {
 	// isRoutable denotes whether this service can be accessed from outside the cluster.
 	isRoutable := !fe.IsSurrogate() &&
 		(svcType != loadbalancer.SVCTypeClusterIP || option.Config.ExternalClusterIP)
+	if sessionAffinity && l7lbProxyPort != 0 {
```
Should we fall back instead of dropping the update? Perhaps one mode should take precedence?

Alternatively, if the settings are incompatible, we should delete the service from bpf. Otherwise whether or not traffic is passed is order-dependent.

borkmann replied:
My preference would be to explicitly fail. This combination is not used today and doesn't make sense in general given session affinity has no effect in case all packets are punted to a L7 proxy. So I think fallback I would rather avoid and instead have it visible to the user.

squeed commented on Documentation/network/kubernetes/kubeproxy-free.rst:
```diff
@@ -391,6 +391,39 @@ corresponding ``NodePort`` and ``ClusterIP`` services. If the annotation
+Host Proxy Delegation
+*********************
+
+If a service backend IP for a given service matches the local node IP,
```
nit: does it make sense to replace "a" with "any"?

borkmann replied:
I think that would distort the meaning, it's not "any" but if the selected backend IP. I'll rephrase a bit to make it more clear. Thx for pointing it out!

squeed commented on pkg/service/service.go:
```diff
@@ -832,11 +832,21 @@ func (s *Service) upsertService(params *lb.SVC) (bool, lb.ID, error) {
-		// filtering out all backend IPs which are not a local endpoint.
-		if filterBackends && len(b.NodeName) > 0 && b.NodeName != nodeTypes.GetName() {
-			continue
+		// Local redirect services or services with externalTrafficPolicy=Local
```
It's not clear to me how this is intended to interact with `external / internalTrafficPolicy`. Does `delegate-if-local` require a `xTrafficPolicy`? What happens if only external or internal are set, but not both?

`delegate-if-local` implies that delegation should take place if there is a local backend, otherwise don't delegate. That doesn't match `xTrafficPolicy`, which **drops** traffic if there are no local backends.

borkmann replied:
The behavior is as with any other regular service. By default (Cluster) we push to both local & remote backends if present, with Local we only filter to local ones and if none is local then traffic is dropped. So same semantics apply.

borkmann committed:
bpf: Lift bind restriction to just look at flag

Small change to just look at the flag, no matter whether there is a
non-zero or zero port. For either case we assume there should be no
rejection.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

dylandreimerink commented on bpf/lib/common.h:
```diff
@@ -655,6 +655,7 @@ enum {
 #define DROP_HOST_NOT_READY	-202
 #define DROP_EP_NOT_READY	-203
 #define DROP_NO_EGRESS_IP	-204
+#define DROP_PUNT_PROXY		-205
```
When adding a new drop code we should also update [api/v1/flow/flow.pb](https://github.com/cilium/cilium/blob/452457e397cfe1bf80d8f59f3c84aa40bf0785aa/api/v1/flow/flow.proto#L375), files generated from it, and [pkg/monitor/api/drop.go](https://github.com/cilium/cilium/blob/452457e397cfe1bf80d8f59f3c84aa40bf0785aa/pkg/monitor/api/drop.go#L17)

borkmann replied:
thx fixed

gandro replied:
Would it make sense to add a comment here saying that this is not a real drop reason? We seem to do something similar for `DROP_NAT_NOT_NEEDED`

https://github.com/cilium/cilium/blob/c6f05a5360918e646126d313a20dd9fb1b4db613/bpf/lib/common.h#L626

borkmann replied:
makes sense, done now.

borkmann committed:
cilium: Regen hubble api wrt drop codes

Add DROP_PUNT_PROXY to the flow.proto file and run generate-hubble-api
to auto-regenerate the drop code table.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

gandro commented on bpf/lib/nodeport.h:
```diff
@@ -1260,6 +1260,12 @@ static __always_inline int nodeport_svc_lb6(struct __ctx_buff *ctx,
 			key, tuple, svc, &ct_state_svc,
 			nodeport_xlate6(svc, tuple), ext_err, 0);
+#ifdef ENABLE_L7_LB
+	if (ret == DROP_PUNT_PROXY) {
```
Similar to `NAT_PUNT_TO_STACK`, would it make sense to introduce a `LB_PUNT_TO_STACK` alias?

borkmann replied:
Makes sense, will add an alias & repush, thx!

gandro reviewed (approved):
ACK for Hubble API

borkmann committed:
bpf: Alias DROP_PUNT_PROXY code

Sebastian suggested we could also alias this drop code similar as we
do with NAT_PUNT_TO_STACK. Thus add an alias LB_PUNT_TO_STACK.

Suggested-by: Sebastian Wicki <sebastian@isovalent.com>
Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

julianwiedmann commented on bpf/lib/lb.h:
```diff
@@ -1062,6 +1105,12 @@ static __always_inline int lb6_local(const void *map, struct __ctx_buff *ctx,
 
 	ipv6_addr_copy(&tuple->daddr, &backend->address);
 
+	if (lb6_svc_is_l7_punt_proxy(svc)) {
+		if (__lookup_ip6_endpoint(&backend->address)) {
+			ret = LB_PUNT_TO_STACK;
+			goto drop_err;
+		}
+	}
```
@borkmann :wave: how is this meant to behave when `lb*_local()` is called from `bpf_lxc`?
